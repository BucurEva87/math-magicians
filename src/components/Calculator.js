import React, { PureComponent } from 'react';
import Display from './Display';
import Button from './Button';
import '../styles/Calculator.scss';
import calculate from '../logic/calculate';

export default class Calculator extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      total: null,
      next: null,
      display: '0',
    };

    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(e) {
    const buttonName = e.target.textContent;

    this.setState((prevState) => ({ ...calculate(prevState, buttonName) }), this.displayOnScreen);
  }

  displayOnScreen() {
    const { total, next } = this.state;

    if (!next && !total) this.setState({ display: '0' });
    else this.setState({ display: next ?? total });
  }

  render() {
    const { display } = this.state;

    return (
      <div id="calculator">
        <Display>{display}</Display>
        <Button id="allClear" handleClick={this.handleClick}>AC</Button>
        <Button id="switchSign" handleClick={this.handleClick}>+/-</Button>
        <Button id="percent" handleClick={this.handleClick}>%</Button>
        <Button id="Buttonide" handleClick={this.handleClick} className="operation">รท</Button>
        <Button id="btn7" handleClick={this.handleClick}>7</Button>
        <Button id="btn8" handleClick={this.handleClick}>8</Button>
        <Button id="btn9" handleClick={this.handleClick}>9</Button>
        <Button id="times" handleClick={this.handleClick} className="operation">x</Button>
        <Button id="btn6" handleClick={this.handleClick}>6</Button>
        <Button id="btn5" handleClick={this.handleClick}>5</Button>
        <Button id="btn4" handleClick={this.handleClick}>4</Button>
        <Button id="minus" handleClick={this.handleClick} className="operation">-</Button>
        <Button id="btn3" handleClick={this.handleClick}>3</Button>
        <Button id="btn2" handleClick={this.handleClick}>2</Button>
        <Button id="btn1" handleClick={this.handleClick}>1</Button>
        <Button id="plus" handleClick={this.handleClick} className="operation">+</Button>
        <Button id="btn0" handleClick={this.handleClick}>0</Button>
        <Button id="dot" handleClick={this.handleClick}>.</Button>
        <Button id="equals" handleClick={this.handleClick} className="operation">=</Button>
      </div>
    );
  }
}
