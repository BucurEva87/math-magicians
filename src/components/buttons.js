export default [
  { id: 'allClear', text: 'AC' },
  { id: 'switchSign', text: '+/-' },
  { id: 'percent', text: '%' },
  { id: 'divide', text: 'รท', className: 'operation' },
  { id: 'btn7', text: '7' },
  { id: 'btn8', text: '8' },
  { id: 'btn9', text: '9' },
  { id: 'times', text: 'x', className: 'operation' },
  { id: 'btn4', text: '4' },
  { id: 'btn5', text: '5' },
  { id: 'btn6', text: '6' },
  { id: 'minus', text: '-', className: 'operation' },
  { id: 'btn1', text: '1' },
  { id: 'btn2', text: '2' },
  { id: 'btn3', text: '3' },
  { id: 'plus', text: '+', className: 'operation' },
  { id: 'btn0', text: '0' },
  { id: 'dot', text: '.' },
  { id: 'equals', text: '=', className: 'operation' },
];
