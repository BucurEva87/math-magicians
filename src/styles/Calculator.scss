@import '../variables/variables.scss';

html {
  font-size: 10px;
}

body {
  width: 100vw;
  height: 100vh;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: $mainFont;
}

#calculator {
  display: grid;
  grid-template-areas:
    'result   result     result  result'
    'allClear switchSign percent divide'
    'btn7     btn8       btn9    times'
    'btn6     btn5       btn4    minus'
    'btn3     btn2       btn1    plus'
    'btn0     btn0       dot     equals';
  width: 60vw;
  background-color: $grayMedium;
  font-size: 3rem;
  gap: 0.3rem;

  #result {
    grid-area: result;
    color: $white;
    background-color: $grayDark;
    display: flex;
    justify-content: flex-end;
    padding: 4rem 1rem;
    font-size: 7rem;
  }

  #allClear {
    grid-area: allClear;
  }

  #switchSign {
    grid-area: switchSign;
  }

  #percent {
    grid-area: percent;
  }

  #divide {
    grid-area: divide;
  }

  #btn7 {
    grid-area: btn7;
  }

  #btn8 {
    grid-area: btn8;
  }

  #btn9 {
    grid-area: btn9;
  }

  #times {
    grid-area: times;
  }

  #btn4 {
    grid-area: btn4;
  }

  #btn5 {
    grid-area: btn5;
  }

  #btn6 {
    grid-area: btn6;
  }

  #minus {
    grid-area: minus;
  }

  #btn1 {
    grid-area: btn1;
  }

  #btn2 {
    grid-area: btn2;
  }

  #btn3 {
    grid-area: btn3;
  }

  #plus {
    grid-area: plus;
  }

  #btn0 {
    grid-area: btn0;
  }

  #dot {
    grid-area: dot;
  }

  #equals {
    grid-area: equals;
  }

  div {
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: $grayLight;
    padding: 5rem;
  }

  .operation {
    background-color: $orange;
  }
}
