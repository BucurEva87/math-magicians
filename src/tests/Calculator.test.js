import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import Calculator from '../components/Calculator';
import buttons from '../components/buttons';

let getByTestId;
let display;
let btn0;
let btn4;
let divide;
let times;
let plus;
let minus;
let equals;
let switchSign;
let dot;

beforeEach(() => {
  const calculator = render(<Calculator />);

  getByTestId = calculator.getByTestId;
  display = getByTestId('display');
  btn0 = getByTestId('btn0');
  btn4 = getByTestId('btn4');
  divide = getByTestId('divide');
  times = getByTestId('times');
  plus = getByTestId('plus');
  minus = getByTestId('minus');
  equals = getByTestId('equals');
  switchSign = getByTestId('switchSign');
  dot = getByTestId('dot');
});

describe('Calculator', () => {
  test('renders on the screen', () => {
    expect(getByTestId('calculator')).toBeInTheDocument();
  });

  test('buttons clicks update the display', () => {
    buttons.forEach((button) => {
      const { id, text } = button;

      if (!/\d/.test(id)) return;

      expect(display).toHaveTextContent('0');

      fireEvent.click(getByTestId(id));

      expect(display).toHaveTextContent(text);

      fireEvent.click(getByTestId('allClear'));
    });
  });

  test('pressing two numerical buttons in a row correctly update the display', () => {
    const btn9 = getByTestId('btn9');

    expect(display).toHaveTextContent('0');

    fireEvent.click(btn4);
    fireEvent.click(btn9);

    expect(display).toHaveTextContent('49');
  });

  test('pressing +/- button switches signs when display has other value than 0', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(btn4);

    expect(display).toHaveTextContent('4');

    fireEvent.click(switchSign);

    expect(display).toHaveTextContent('-4');
  });

  test('pressing +/- button does not switch signs when display has a value of 0', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(switchSign);

    expect(display).toHaveTextContent('0');
  });

  test('pressing +/- button twice does not switch the sign', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(btn4);

    expect(display).toHaveTextContent('4');

    fireEvent.click(switchSign);
    fireEvent.click(switchSign);

    expect(display).toHaveTextContent('4');
  });

  test('pressing the dot button adds a decimal sign at the end of the display', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(dot);

    expect(display).toHaveTextContent('0.');
  });

  test('pressing the dot button multiple times adds a single decimal sign at the end of the display', () => {
    expect(display).toHaveTextContent('0');

    for (let i = 0; i < 5; i += 1) fireEvent.click(dot);

    expect(display).toHaveTextContent('0.');
  });

  test('pressing the 0 button has no impact if the displayed value is already 0', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(btn0);

    expect(display).toHaveTextContent('0');
  });

  test('trying to divide by zero returns a math violation inside the display', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(btn4);

    expect(display).toHaveTextContent('4');

    fireEvent.click(divide);

    expect(display).toHaveTextContent('4');

    fireEvent.click(btn0);

    expect(display).toHaveTextContent('0');

    fireEvent.click(equals);

    expect(display).toHaveTextContent('Can\'t divide by 0.');
  });

  test('pressing divide button correctly divides the two operands', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(btn4);
    fireEvent.click(btn0);

    expect(display).toHaveTextContent('40');

    fireEvent.click(divide);
    fireEvent.click(btn4);
    fireEvent.click(equals);

    expect(display).toHaveTextContent('10');
  });

  test('pressing times button correctly multiplies the two operands', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(btn4);

    expect(display).toHaveTextContent('4');

    fireEvent.click(times);
    fireEvent.click(btn4);
    fireEvent.click(equals);

    expect(display).toHaveTextContent('16');
  });

  test('pressing times button correctly multiplies two operands of negative signs', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(btn4);
    fireEvent.click(switchSign);

    expect(display).toHaveTextContent('-4');

    fireEvent.click(times);
    fireEvent.click(btn4);
    fireEvent.click(switchSign);
    fireEvent.click(equals);

    expect(display).toHaveTextContent('16');
  });

  test('pressing times button correctly multiplies two operands of mixed signs', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(btn4);
    fireEvent.click(switchSign);

    expect(display).toHaveTextContent('-4');

    fireEvent.click(times);
    fireEvent.click(btn4);
    fireEvent.click(equals);

    expect(display).toHaveTextContent('-16');
  });

  test('pressing plus button correctly adds the two operands', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(btn4);

    expect(display).toHaveTextContent('4');

    fireEvent.click(plus);
    fireEvent.click(btn4);
    fireEvent.click(equals);

    expect(display).toHaveTextContent('8');
  });

  test('pressing minus button correctly subtracts the second operand from the first', () => {
    expect(display).toHaveTextContent('0');

    fireEvent.click(btn0);
    fireEvent.click(minus);
    fireEvent.click(btn4);
    fireEvent.click(equals);

    expect(display).toHaveTextContent('-4');
  });

  test('executing multiple operations in a row works correctly', () => {
    const btn7 = getByTestId('btn7');
    const btn3 = getByTestId('btn3');

    expect(display).toHaveTextContent('0');

    fireEvent.click(btn7);
    fireEvent.click(times);
    fireEvent.click(btn3);
    fireEvent.click(equals);

    expect(display).toHaveTextContent('21');

    fireEvent.click(divide);
    fireEvent.click(btn4);
    fireEvent.click(equals);

    expect(display).toHaveTextContent('5.25');
  });

  test('matches snapshot', () => {
    const tree = renderer.create(<Calculator />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
