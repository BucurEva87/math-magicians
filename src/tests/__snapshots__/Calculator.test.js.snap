// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator matches snapshot 1`] = `
<div
  data-testid="calculator"
  id="calculator"
>
  <div
    data-testid="display"
    id="result"
  >
    0
  </div>
  <button
    className=""
    data-testid="allClear"
    id="allClear"
    onClick={[Function]}
    type="button"
  >
    AC
  </button>
  <button
    className=""
    data-testid="switchSign"
    id="switchSign"
    onClick={[Function]}
    type="button"
  >
    +/-
  </button>
  <button
    className=""
    data-testid="percent"
    id="percent"
    onClick={[Function]}
    type="button"
  >
    %
  </button>
  <button
    className="operation"
    data-testid="divide"
    id="divide"
    onClick={[Function]}
    type="button"
  >
    รท
  </button>
  <button
    className=""
    data-testid="btn7"
    id="btn7"
    onClick={[Function]}
    type="button"
  >
    7
  </button>
  <button
    className=""
    data-testid="btn8"
    id="btn8"
    onClick={[Function]}
    type="button"
  >
    8
  </button>
  <button
    className=""
    data-testid="btn9"
    id="btn9"
    onClick={[Function]}
    type="button"
  >
    9
  </button>
  <button
    className="operation"
    data-testid="times"
    id="times"
    onClick={[Function]}
    type="button"
  >
    x
  </button>
  <button
    className=""
    data-testid="btn4"
    id="btn4"
    onClick={[Function]}
    type="button"
  >
    4
  </button>
  <button
    className=""
    data-testid="btn5"
    id="btn5"
    onClick={[Function]}
    type="button"
  >
    5
  </button>
  <button
    className=""
    data-testid="btn6"
    id="btn6"
    onClick={[Function]}
    type="button"
  >
    6
  </button>
  <button
    className="operation"
    data-testid="minus"
    id="minus"
    onClick={[Function]}
    type="button"
  >
    -
  </button>
  <button
    className=""
    data-testid="btn1"
    id="btn1"
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    className=""
    data-testid="btn2"
    id="btn2"
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    className=""
    data-testid="btn3"
    id="btn3"
    onClick={[Function]}
    type="button"
  >
    3
  </button>
  <button
    className="operation"
    data-testid="plus"
    id="plus"
    onClick={[Function]}
    type="button"
  >
    +
  </button>
  <button
    className=""
    data-testid="btn0"
    id="btn0"
    onClick={[Function]}
    type="button"
  >
    0
  </button>
  <button
    className=""
    data-testid="dot"
    id="dot"
    onClick={[Function]}
    type="button"
  >
    .
  </button>
  <button
    className="operation"
    data-testid="equals"
    id="equals"
    onClick={[Function]}
    type="button"
  >
    =
  </button>
</div>
`;
